File: A2Z_striver/patterns/pattern_6/pattern_6.cpp
#include <iostream>

using namespace std;

// Header Comment Block
// Problem: Print an Inverted Number Triangle.
// Example Input: n = 5
// Example Output:
// 1 2 3 4 5
// 1 2 3 4
// 1 2 3
// 1 2
// 1
// Goal: O(N^2) Time | O(1) Space

class Solution
{
public:
    // ~ Approach 1: 1-Based Indexing (Your Approach)
    // Analysis: Time O(N^2) | Space O(1)
    void pattern6_v1(int n)
    {
        for (int i = 1; i <= n; i++)
        {
            // ! Math check: n - i + 1
            // i=1 -> limit 5
            // i=2 -> limit 4
            for (int j = 1; j <= n - i + 1; j++)
            {
                cout << j << " ";
            }
            cout << "\n";
        }
    }

    // ~ Approach 2: 0-Based Indexing (Standard Loop)
    // Analysis: Time O(N^2) | Space O(1)
    void pattern6_v2(int n)
    {
        for (int i = 0; i < n; i++)
        {
            // ! Math check: limit is simply n - i
            for (int j = 0; j < n - i; j++)
            {
                cout << j + 1 << " ";
            }
            cout << "\n";
        }
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;
    int n = 5;

    cout << "--- Approach 1 (1-based) ---\n";
    sol.pattern6_v1(n);

    cout << "\n--- Approach 2 (0-based) ---\n";
    sol.pattern6_v2(n);

    return 0;
}

File: A2Z_striver/1 Basics/1.2 Patterns/pattern_6/pattern_6.py
import sys

class Solution:
    """
    Problem: Print an Inverted Number Triangle.
    Example Input: n = 5
    Example Output:
    1 2 3 4 5
    1 2 3 4
    1 2 3
    1 2
    1
    
    Constraints: 1 <= n <= 100
    Goal: O(N^2) Time.
    """

    # // ~ Approach 1: Nested Loops (Standard Algo)
    # // Analysis: Time O(N^2) | Space O(1)
    def pattern6_nested(self, n):
        for i in range(n):
            # // ! Logic: Run loop (n-i) times
            # // - We range from 1 to (n-i) inclusive for cleaner printing
            for j in range(1, n - i + 1):
                print(j, end=" ")
            print()

    # // ~ Approach 2: Pythonic (List Comprehension)
    # // Analysis: Time O(N^2) | Space O(N) (Temp list creation)
    def pattern6_pythonic(self, n):
        for i in range(n):
            # // . Generate list of numbers, map to str, and join
            print(" ".join(str(x) for x in range(1, n - i + 1)))

# Driver Code
if __name__ == "__main__":
    solver = Solution()
    N = 5
    
    print(f"--- Approach 1: Nested Loops (N={N}) ---")
    solver.pattern6_nested(N)
    
    print(f"\n--- Approach 2: Pythonic (N={N}) ---")
    solver.pattern6_pythonic(N)